src/eigen/linalg.jl

```diff
@@ -0,0 +1,19 @@
+# Calculate the norms of the columns of an array
+function columnwise_norms(X::AbstractArray{T}) where{T}
+    vec(sqrt.(sum(abs2, X; dims=1)))
+end
+
+# Returns a vector of dot(A[:, i], B[:, i]), for all columns of A, B
+@views function diag_prod(A::AbstractArray{T}, B::AbstractArray{T}) where {T}
+    @assert size(A) == size(B)
+    [real(dot(A[:, i], B[:, i])) for i = 1:size(A, 2)]
+end
+
+# Returns a vector of dot(A[:, i], Diagonal(diag), B[:, i]), for all columns of
+# A, B, and a vector diag
+@views function diag_prod(A::AbstractArray{T}, diag::AbstractVector,
```

Comment:
Why not explicitly have `diag::Diagonal` and than pass it through to `dot` ? Or rather here it could even be D::AbstractMatrix without much changes, right ? (May not be so easy for the GPU implementation, though)

Reply:
I could probably pass an `AbstractMatrix` to the function, and have a special case with a `isdiag()` test on the GPU (and otherwise do the full MM). Not sure if it is worth the extra code.

Reply:
Just specialise on the type ::Diagonal for GPU ? From the diagonal object you can extract the actual vector as a GPU array